packages/@headlessui-react/src/components/combobox/combobox.tsx

```diff
@@ -1424,16 +1453,31 @@ function ButtonFn<TTag extends ElementType = typeof DEFAULT_BUTTON_TAG>(
     }
   })
 
-  let handleClick = useEvent((event: ReactMouseEvent<HTMLButtonElement>) => {
-    if (isDisabledReactIssue7711(event.currentTarget)) return event.preventDefault()
-    if (data.comboboxState === ComboboxState.Open) {
-      actions.closeCombobox()
-    } else {
-      event.preventDefault()
-      actions.openCombobox()
+  let handleMouseDown = useEvent((event: ReactMouseEvent<HTMLButtonElement>) => {
+    // We use the `mousedown` event because this will fire before the `focus`
+    // event. When we use `event.preventDefault()` here, the
+    // `document.activeElement` will stay on the current element.
+    //
+    // Typically that means that the `ComboboxInput` will stay focused, and thus
+    // the selection / cursor position will be preserved.
+    event.preventDefault()
+
+    if (isDisabledReactIssue7711(event.currentTarget)) return
+
+    // When we used the `click` event we didn't have to worry about this because
+    // that only fires if you use the `left` mouse button. However, now that
+    // we use the `mousedown` event, we do have to worry about which button is
+    // pressed.
```

Comment:
I think it would be better to write this comment so that it explains the current behavior, not the past implementation, since that probably just makes things more confusing for future readers.

Something like this:

```suggestion
    // Since we're using the `mousedown` event instead of a `click` event
    // here to preserve the focus of the `ComboboxInput`, we need to also
    // check that the `left` mouse button was clicked.
```